data/abi/abi_value.go

```diff
@@ -0,0 +1,304 @@
+		value:   values,
+	}, nil
+}
+
+// MakeTuple takes an array of ABI values and returns an ABI tuple value.
+func MakeTuple(values []Value) (Value, error) {
+	if len(values) >= (1 << 16) {
+		return Value{}, fmt.Errorf("tuple make error: pass in argument number larger than 2^16")
+	}
+	tupleType := make([]Type, len(values))
+	for i := 0; i < len(values); i++ {
+		tupleType[i] = values[i].ABIType
+	}
+
+	castedTupleType, err := MakeTupleType(tupleType)
+	if err != nil {
+		return Value{}, err
+	}
+
+	return Value{
+		ABIType: castedTupleType,
+		value:   values,
+	}, nil
+}
+
+// MakeBool takes a boolean value and returns an ABI bool value.
+func MakeBool(value bool) Value {
+	return Value{
+		ABIType: MakeBoolType(),
+		value:   value,
+	}
+}
+
+// GetUint8 tries to retreve an uint8 from an ABI Value.
+func (v Value) GetUint8() (uint8, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 8 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint8(bigIntForm.Uint64()), nil
+}
+
+// GetUint16 tries to retrieve an uint16 from an ABI Value.
+func (v Value) GetUint16() (uint16, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 16 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint16(bigIntForm.Uint64()), nil
+}
+
+// GetUint32 tries to retrieve an uint32 from an ABI Value.
+func (v Value) GetUint32() (uint32, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 32 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return uint32(bigIntForm.Uint64()), nil
+}
+
+// GetUint64 tries to retrieve an uint64 from an ABI Value.
+func (v Value) GetUint64() (uint64, error) {
+	if v.ABIType.abiTypeID != Uint || v.ABIType.bitSize > 64 {
+		return 0, fmt.Errorf("value type unmatch or bitSize too large")
+	}
+	bigIntForm, err := v.GetUint()
+	if err != nil {
+		return 0, err
+	}
+	return bigIntForm.Uint64(), nil
+}
+
+// GetUint tries to retrieve an big uint from an ABI Value.
+func (v Value) GetUint() (*big.Int, error) {
+	if v.ABIType.abiTypeID != Uint {
+		return nil, fmt.Errorf("value type unmatch")
+	}
+	bigIntForm := v.value.(*big.Int)
+	sizeThreshold := big.NewInt(0).Lsh(big.NewInt(1), uint(v.ABIType.bitSize))
+	if sizeThreshold.Cmp(bigIntForm) <= 0 {
+		return nil, fmt.Errorf("value is larger than uint bitSize")
+	}
+	return bigIntForm, nil
+}
+
+// GetUfixed tries to retrieve an big rational number from an ABI Value.
+func (v Value) GetUfixed() (*big.Int, error) {
+	if v.ABIType.abiTypeID != Ufixed {
+		return nil, fmt.Errorf("value type unmatch, should be ufixed")
+	}
+	bigIntForm := v.value.(*big.Int)
+	sizeThreshold := big.NewInt(0).Lsh(big.NewInt(1), uint(v.ABIType.bitSize))
+	if sizeThreshold.Cmp(bigIntForm) <= 0 {
```

Comment:
The comment says it returns a big.Rat, but I think it just returns the value, scaled up to an integer?

Reply:
Should it return the width as well, then?  It seems weird to "lose information" when using these Get* accessors.  I suppose the caller could read it from the value (is the width a public field?)